### ISSUE TITLE
"nuget.exe push" reads the package file into memory even if the transaction is canceled

### ISSUE THREAD

**Issue by spadapet** [NuGet/Home#2430]
@zarenner showed me this: He was pushing a giant (6GB) package to visualstudio.com that fails authorization, yet the whole 6GB file was streamed into memory even though it isn't sent. That takes 40 seconds when the HTTP transaction immediately fails.

Related to: https://github.com/NuGet/Home/issues/1501

We should find out why the file is being read into memory and try to prevent it when the HTTP transaction fails.

**Comment by yishaigalatzer** [comment]
@spadapet is this after your fix to use chunked encoding?

**Comment by zarenner** [comment]
@yishaigalatzer Yes, this is with Peter's fix.

**Comment by spadapet** [comment]
That's right, this happens with chunking. At least it's not sent over the wire, but I need to see why the file is even being read.

**Comment by spadapet** [comment]
Looks like the file will always be read due to code in:
  System.Net.Http.HttpClientHandler.GetRequestStreamCallback
That code doesn't know that the HTTP request was unauthorized and will always copy the content stream to the output.

The only way to avoid this might be to trigger some other request before the PUT so that authentication happens. That's a quick fix. A long term fix would be to ask the server if a PUT request would succeed given specific package metadata. That way the whole nupack wouldn't need to be sent to the server just to find out that it would fail (only the metadata)

**Comment by spadapet** [comment]
Save this here: https://blogs.msdn.microsoft.com/ncl/2009/05/05/custom-http-authentication-schemes/

**Comment by spadapet** [comment]
That link was from Chris Ross (chrross). The suggestion is to implement a custom authentication manager to provide credentials so that the HTTP request doesn't fail. Then we don't need to try again with credentials.
